### PR TITLE
Laboratory work #2, Ksenia Maiorova - 23FPL2

### PR DESCRIPTION
**Pull request by ksenia-maiorova**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,6 +20,9 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+    return re.sub(r'[^\s\w]+|\d+', r' ', text.lower()).split()
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:25]: is it working? can you explain this regular expression

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -62,6 +86,16 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
...
+    if (not vocab or not document_tokens or not isinstance(vocab, list) or
+            not isinstance(document_tokens, list)):
+        return None
+    if (not all(isinstance(term, str) for term in vocab) or
+            not all(isinstance(token, str) for token in document_tokens)):
+        return None
+    for token in document_tokens:
+        if token not in vocab:
+            vocab.append(token)
+    return dict(zip(vocab, [document_tokens.count(term) / len(document_tokens) for term in vocab]))
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:98]: looks good, can you explain what is going here?

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -120,6 +175,25 @@ def calculate_bm25(
...
+        return None
+    if (isinstance(doc_len, bool) or not all(isinstance(term, str) for term in vocab) or
+            not all(isinstance(term, str) for term in idf_document)):
+        return None
+    if (not all(isinstance(term, str) for term in document) or
+            not all(isinstance(freq, float) for freq in idf_document.values())):
+        return None
+    vocab.extend([term for term in document if term not in vocab])
+    idf_document.update({term: 0.0 for term in vocab if term not in idf_document})
+    return {term: (idf_document[term] * ((document.count(term) * (k1 + 1))
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:193]: looks massive, place the code more beautifully

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -138,6 +212,27 @@ def rank_documents(
...
+            not all(isinstance(index, dict) for index in indexes)):
+        return None
+    tokenized_query = tokenize(query)
+    if not tokenized_query:
+        return None
+    clean_query = remove_stopwords(tokenized_query, stopwords)
+    if not clean_query:
+        return None
+    index_score_list = []
+    for index in indexes:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:228]: try to use enumerate

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -138,6 +212,27 @@ def rank_documents(
...
+        metric_score = 0.0
+        index.update({term: 0.0 for term in clean_query if term not in index})
+        for term in clean_query:
+            metric_score += index[term]
+        index_score = (indexes.index(index), metric_score)
+        index_score_list.append(index_score)
+    return sorted(index_score_list, key=lambda tple: tple[-1], reverse=True)
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:235]: rename tple

File: lab_2_retrieval_w_bm25/start.py [hunk 6]
```diff
@@ -26,7 +29,45 @@ def main() -> None:
...
+        bm25_list.append(bm25)
+    query = "Which fairy tale has Fairy Queen?"
+    tfidf_ranking = rank_documents(tf_idf_list, query, stopwords)
+    bm25_ranking = rank_documents(bm25_list, query, stopwords)
+    if not tfidf_ranking or not bm25_ranking:
+        return None
+    print(f"{tfidf_ranking}\n{bm25_ranking}")
+    result = [index_score_1 for index_score_1, index_score_2 in zip(tfidf_ranking, bm25_ranking)
+              if index_score_1 == index_score_2]
+    print(result)
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:70]: is it working?

**Review by EIzharikov — COMMENTED** [review]
A great start! Don't slow down and get ten! Good luck!

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (ksenia-maiorova) [reply lab_2_retrieval_w_bm25/main.py:25]: yes, it is working... 
this regular expression filters out symbols that are not a whitespace character or a letter. then they are replaced with a whitespace for each symbol in lowered text
[^] - everything but the expression in the brackets
\s - only whitespace characters
\w - only letters, digits and _
+ - one or more repetitions
| - or 
\d - only digits

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (ksenia-maiorova) [reply lab_2_retrieval_w_bm25/main.py:235]: fixed

File: lab_2_retrieval_w_bm25/start.py (hunk 6 above)
Comment (ksenia-maiorova) [reply lab_2_retrieval_w_bm25/start.py:70]: yep
should i explain anything?

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (ksenia-maiorova) [reply lab_2_retrieval_w_bm25/main.py:193]: fixed (??)
hopefully

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (ksenia-maiorova) [reply lab_2_retrieval_w_bm25/main.py]: tried

File: lab_2_retrieval_w_bm25/start.py [hunk 7]
```diff
@@ -29,44 +29,34 @@ def main() -> None:
...
     clean_documents = []
     for document in documents:
         tokenized_document = tokenize(document)
-        if not tokenized_document:
-            return None
-        clean_document = remove_stopwords(tokenized_document, stopwords)
-        if not clean_document:
-            return None
-        clean_documents.append(clean_document)
+        if tokenized_document:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:36]: reverse statement, retrun None early

File: lab_2_retrieval_w_bm25/start.py [hunk 8]
```diff
@@ -29,44 +29,34 @@ def main() -> None:
...
+            clean_document = remove_stopwords(tokenized_document, stopwords)
+            if clean_document:
+                clean_documents.append(clean_document)
     vocab = build_vocabulary(clean_documents)
     if not vocab:
-        return None
-    idf = calculate_idf(vocab, clean_documents)
-    if not idf:
-        return None
+        idf = calculate_idf(vocab, clean_documents)
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:42]: why it is after ```if not vocab:```  ?

File: lab_2_retrieval_w_bm25/start.py [hunk 9]
```diff
@@ -29,44 +29,34 @@ def main() -> None:
...
     bm25_list = []
     for doc in clean_documents:
         tf = calculate_tf(vocab, doc)
-        if not tf:
-            return None
-        tf_idf = calculate_tf_idf(tf, idf)
-        if not tf_idf:
-            return None
-        tf_idf_list.append(tf_idf)
+        if tf:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:48]: reverse if statement

**Review by EIzharikov — COMMENTED** [review]
You've made some progress, but it's still progress! I wish you good luck in further work!

File: lab_2_retrieval_w_bm25/start.py (hunk 8 above)
Comment (ksenia-maiorova) [reply lab_2_retrieval_w_bm25/start.py:42]: actually i dont know why its there...
fixed

File: lab_2_retrieval_w_bm25/start.py (hunk 7 above)
Comment (ksenia-maiorova) [reply lab_2_retrieval_w_bm25/start.py:36]: done

File: lab_2_retrieval_w_bm25/start.py (hunk 9 above)
Comment (ksenia-maiorova) [reply lab_2_retrieval_w_bm25/start.py:48]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py]: and what??

**Review by EIzharikov — COMMENTED** [review]
Make CI green, go for 10 and be ready for the next review. Good luck!

**Review by EIzharikov — COMMENTED** [review]
Make CI green and be ready for Friday. Good luck.

**Pull request closed without merging**